include/zephyr/arch/arch_interface.h

```diff
@@ -1240,6 +1240,35 @@ bool arch_pcie_msi_vector_connect(msi_vector_t *vector,
  */
 void arch_spin_relax(void);
 
+#ifdef CONFIG_ARCH_STACKWALK
+/**
+ * stack_trace_consume_fn - Callback for @ref arch_stack_walk
+ * @param cookie Caller supplied pointer handed back by @ref arch_stack_walk
+ * @param addr The stack entry address to consume
+ *
+ * @return True, if the entry was consumed or skipped. False, if there is no space left to store
+ */
+typedef bool (*stack_trace_consume_fn)(void *cookie, unsigned long addr);
+
+/**
+ * @brief Architecture-specific function to walk the stack
+ *
+ * @param consume_entry Callback which is invoked by the architecture code for each entry.
+ * @param cookie Caller supplied pointer which is handed back to @a consume_entry
+ * @param csf Pointer to a callee-saved-registers struct, can be NULL
+ * @param esf Pointer to esf struct, can be NULL
+ *
+ * ============ ======= ============================================
+ * csf	        esf
+ * ============ ======= ============================================
+ * csf		NULL	Stack trace from csf (can be current)
+ * csf		esf	Stack trace starting on esf
+ * ============ ======= ============================================
+ */
+void arch_stack_walk(stack_trace_consume_fn consume_entry, void *cookie,
+		     const void *csf, const void *esf);
```

Comment:
I'm not sure if this is the right place for this function, not sure if `csf` & `esf` are too low level. They are `const void *` because their type isn't (probably can't be?) available in this file.